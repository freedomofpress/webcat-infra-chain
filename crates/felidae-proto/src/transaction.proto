syntax = "proto3";

package felidae.transaction;

message Transaction {
    string chain_id = 1;
}