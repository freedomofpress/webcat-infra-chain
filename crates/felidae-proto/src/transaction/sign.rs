use std::{any::Any, collections::HashMap};

use prost::{Message as _, bytes::Bytes};
pub use ring::error::Unspecified as VerifyError;
use ring::{
    digest::{Context, Digest},
    signature::{ECDSA_P256_SHA256_FIXED, UnparsedPublicKey},
};

use felidae_traverse::Traverse;

mod signer;
pub use signer::{AsyncSigner, KeyPair, KeyPairs, Signer};

#[derive(thiserror::Error, Debug)]
pub enum SignError {
    #[error("No keypair available for public key in transaction")]
    MissingKeypair,
    #[error("Transaction already contains a signature")]
    AlreadySigned,
}

impl super::Signature {
    /// Create a new blank signature for the given public key.
    pub fn unsigned(public_key: Bytes) -> Self {
        Self {
            public_key,
            signature: Bytes::new(),
        }
    }
}

impl From<super::Signature> for UnparsedPublicKey<Bytes> {
    fn from(sig: super::Signature) -> Self {
        UnparsedPublicKey::new(&ECDSA_P256_SHA256_FIXED, sig.public_key)
    }
}

impl super::Transaction {
    /// Decode a transaction and verify all its signatures, stripping them in the process.
    pub fn authenticate_from_proto<B: AsRef<[u8]>>(
        context: Context,
        buf: B,
    ) -> Result<Self, VerifyError> {
        let tx = Self::decode(buf.as_ref()).map_err(|_| VerifyError)?;
        let mut tx = tx.verify_all(context).map_err(|_| VerifyError)?;
        tx.unsign_all();
        Ok(tx)
    }

    /// Deserialize a transaction from JSON and verify all its signatures, stripping them in the process.
    pub fn authenticate_from_json(context: Context, s: &str) -> Result<Self, VerifyError> {
        let tx: Self = serde_json::from_str(s).map_err(|_| VerifyError)?;
        let mut tx = tx.verify_all(context).map_err(|_| VerifyError)?;
        tx.unsign_all();
        Ok(tx)
    }

    /// Encode a transaction and fill in all its signatures using the given signer.
    pub fn sign_to_proto(
        self,
        context: Context,
        signer: impl Signer,
    ) -> Result<Vec<u8>, SignError> {
        let tx = self.sign_all(signer, context)?;
        Ok(tx.encode_to_vec())
    }

    /// Serialize a transaction to JSON and fill in all its signatures using the given signer.
    pub fn sign_to_json(self, context: Context, signer: impl Signer) -> Result<String, SignError> {
        let tx = self.sign_all(signer, context)?;
        serde_json::to_string(&tx).map_err(|_| SignError::AlreadySigned)
    }

    /// Compute the hash of the transaction as a protobuf message.
    ///
    /// The hash is computed after removing all signatures from the transaction, so that it can
    /// be computed as an *input* to signing.
    pub fn hash(&self, mut context: Context) -> Digest
    where
        Self: prost::Message + Sized,
    {
        let mut unbound = self.clone();
        unbound.unsign_all();
        context.update(&unbound.encode_to_vec());
        context.finish()
    }

    /// Fill in every blank signature with a valid signature over the hash of the transaction.
    fn sign_all(mut self, signer: impl Signer, context: Context) -> Result<Self, SignError>
    where
        Self: prost::Message + Sized,
    {
        // Compute the transaction hash once:
        let digest = self.hash(context);

        // Then traverse the transaction, filling in signatures as we go:
        let mut result = Ok(());
        self.traverse_mut(&mut |v| {
            if let Some(super::Signature {
                public_key,
                signature,
            }) = (v as &mut dyn Any).downcast_mut::<super::Signature>()
            {
                if signature.is_empty() {
                    if let Some(sig) = signer.sign_with(public_key.as_ref(), digest) {
                        *signature = sig.into();
                    } else {
                        // No keypair available:
                        result = Err(SignError::MissingKeypair);
                    }
                } else {
                    // Signature already present:
                    result = Err(SignError::AlreadySigned);
                }
            }
        });

        // If we succeeded, return the signed transaction:
        result.map(|()| self)
    }

    /// Remove all signatures from this object and its sub-objects.
    fn unsign_all(&mut self) {
        self.traverse_mut(&mut |v| {
            if let Some(super::Signature { signature, .. }) =
                (v as &mut dyn Any).downcast_mut::<super::Signature>()
            {
                signature.clear();
            }
        });
    }

    /// Verify every signature in the transaction against the hash of the transaction.
    fn verify_all(self, context: Context) -> Result<Self, VerifyError> {
        let digest = self.hash(context);
        let mut result = Ok(());
        self.traverse(&mut |v| {
            if let Some(super::Signature {
                public_key,
                signature,
                ..
            }) = (v as &dyn Any).downcast_ref::<super::Signature>()
            {
                if !signature.is_empty() {
                    if let Err(e) = UnparsedPublicKey::new(&ECDSA_P256_SHA256_FIXED, public_key)
                        .verify(digest.as_ref(), signature)
                    {
                        result = Err(e);
                    }
                } else {
                    // Missing signature is an error:
                    result = Err(VerifyError);
                }
            }
        });
        result.map(|()| self)
    }
}

mod test;
