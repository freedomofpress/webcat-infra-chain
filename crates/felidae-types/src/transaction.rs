use aws_lc_rs::digest::{Context, SHA256};
use felidae_proto::transaction::{self as proto};
use prost::bytes::Bytes;
use std::{hash::Hash, ops::Deref};
use tendermint::AppHash;

use crate::{SignError, Signer};

/// Type conversions between the protobuf-generated types and the domain types.
mod convert;

mod authenticated;
pub use authenticated::AuthenticatedTx;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Transaction {
    pub chain_id: String,
    pub actions: Vec<Action>,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Unsigned {
    pub public_key: Bytes,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Action {
    Reconfigure(Reconfigure),
    Observe(Observe),
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Reconfigure {
    pub admin: Admin,
    pub config: Config,
    pub version: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Config {
    pub admin_config: AdminConfig,
    pub oracle_config: OracleConfig,
    pub onion_config: OnionConfig,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct AdminConfig {
    pub admins: Vec<Admin>,
    pub voting_config: VotingConfig,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Admin {
    identity: Bytes,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct OracleConfig {
    pub enabled: bool,
    pub oracles: Vec<Oracle>,
    pub voting_config: VotingConfig,
    pub max_enrolled_subdomains: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Oracle {
    identity: Bytes,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct OnionConfig {
    pub enabled: bool,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct VotingConfig {
    pub total: Total,
    pub quorum: Quorum,
    pub timeout: Timeout,
    pub delay: Delay,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Total(pub u64);

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Quorum(pub u64);

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Timeout(pub u64);

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Delay(pub u64);

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Observe {
    pub oracle: Oracle,
    pub observation: Observation,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Observation {
    pub domain: fqdn::FQDN,
    pub hash_observed: HashObserved,
    pub blockstamp: Blockstamp,
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct HashObserved(pub [u8; 32]);

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Blockstamp {
    pub app_hash: AppHash,
    pub block_number: u64,
}

impl PartialOrd for Blockstamp {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Blockstamp {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.block_number
            .cmp(&other.block_number)
            .then_with(|| self.app_hash.as_bytes().cmp(other.app_hash.as_bytes()))
    }
}

impl Hash for Blockstamp {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.app_hash.as_bytes().hash(state);
        self.block_number.hash(state);
    }
}
