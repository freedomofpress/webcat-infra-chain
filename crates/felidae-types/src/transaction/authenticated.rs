use super::*;

/// A transaction that has been signed and whose signatures have been verified.
///
/// This transaction is not necessarily valid, either internally or against the current state.
/// However, it is guaranteed that it was signed by the claimed public keys.
#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct AuthenticatedTx(Transaction);

impl Deref for AuthenticatedTx {
    type Target = Transaction;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl AuthenticatedTx {
    /// Deserialize a transaction from JSON, verify all its signatures, and convert it into the
    /// domain type.
    pub fn from_json(json: &str) -> Result<AuthenticatedTx, crate::ParseError> {
        Ok(AuthenticatedTx(
            proto::Transaction::authenticate_from_json(json)?.try_into()?,
        ))
    }

    /// Decode a transaction from bytes, verify all its signatures, and convert it into the domain
    /// type.
    pub fn from_proto<B: AsRef<[u8]>>(buf: B) -> Result<AuthenticatedTx, crate::ParseError> {
        Ok(AuthenticatedTx(
            proto::Transaction::authenticate_from_proto(buf)?.try_into()?,
        ))
    }
}

impl Transaction {
    /// Serialize the transaction to JSON, signing all its actions with the given signer.
    pub fn sign_to_json(self, signer: impl Signer) -> Result<String, SignError> {
        proto::Transaction::from(self).sign_to_json(signer)
    }

    /// Encode the transaction to bytes, signing all its actions with the given signer.
    pub fn sign_to_proto(self, signer: impl Signer) -> Result<Vec<u8>, SignError> {
        proto::Transaction::from(self).sign_to_proto(signer)
    }
}
