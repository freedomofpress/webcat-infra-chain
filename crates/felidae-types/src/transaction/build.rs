use prost::bytes::Bytes;
use tendermint::Time;

use crate::transaction::{
    Action, Admin, ChainId, Config, Observation, Observe, Oracle, Reconfigure, Transaction,
};

pub struct Builder {
    chain_id: ChainId,
    actions: Vec<Action>,
}

impl Builder {
    pub fn new(chain_id: ChainId) -> Self {
        Self {
            chain_id,
            actions: Vec::new(),
        }
    }

    fn action(mut self, action: Action) -> Self {
        self.actions.push(action);
        self
    }

    pub fn reconfigure(
        self,
        admin: impl Into<Bytes>,
        not_before: Time,
        not_after: Time,
        config: Config,
    ) -> Self {
        self.action(Action::Reconfigure(Reconfigure {
            admin: Admin {
                identity: admin.into(),
            },
            config,
            not_before,
            not_after,
        }))
    }

    pub fn observe(self, oracle: impl Into<Bytes>, observation: Observation) -> Self {
        self.action(Action::Observe(Observe {
            oracle: Oracle {
                identity: oracle.into(),
            },
            observation,
        }))
    }

    pub fn build(self) -> Transaction {
        Transaction {
            chain_id: self.chain_id,
            actions: self.actions,
        }
    }
}
