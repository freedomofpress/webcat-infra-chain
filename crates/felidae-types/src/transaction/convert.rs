use felidae_proto::transaction::{self as proto, KeyPair};
use fqdn::FQDN;

use super::*;

impl TryFrom<proto::Transaction> for Transaction {
    type Error = crate::ParseError;

    fn try_from(tx: proto::Transaction) -> Result<Self, Self::Error> {
        let proto::Transaction { chain_id, actions } = tx;

        let chain_id = ChainId(chain_id);

        let actions = actions
            .into_iter()
            .map(TryInto::try_into)
            .collect::<Result<_, _>>()?;

        Ok(Transaction { chain_id, actions })
    }
}

impl From<Transaction> for proto::Transaction {
    fn from(tx: Transaction) -> Self {
        let Transaction { chain_id, actions } = tx;
        proto::Transaction {
            chain_id: chain_id.0,
            actions: actions.into_iter().map(Into::into).collect(),
        }
    }
}

impl TryFrom<String> for ChainId {
    type Error = crate::ParseError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        if value.is_empty() {
            Err(crate::ParseError::new::<ChainId>(value))
        } else {
            Ok(ChainId(value))
        }
    }
}

impl From<ChainId> for String {
    fn from(value: ChainId) -> Self {
        value.0
    }
}

impl TryFrom<String> for Domain {
    type Error = crate::ParseError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        let name = FQDN::from_ascii_str(&value)
            .map_err(|_| crate::ParseError::new::<Domain>(value.clone()))?;
        Ok(Domain { name })
    }
}

impl From<Domain> for String {
    fn from(value: Domain) -> Self {
        value.name.to_string()
    }
}

impl TryFrom<String> for Zone {
    type Error = crate::ParseError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        let name = FQDN::from_ascii_str(&value)
            .map_err(|_| crate::ParseError::new::<Zone>(value.clone()))?;
        Ok(Zone { name })
    }
}

impl From<Zone> for String {
    fn from(value: Zone) -> Self {
        value.name.to_string()
    }
}

impl From<PrefixOrderDomain> for String {
    fn from(value: PrefixOrderDomain) -> Self {
        value.name.to_string()
    }
}

impl TryFrom<String> for PrefixOrderDomain {
    type Error = crate::ParseError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        let name = FQDN::from_ascii_str(&value)
            .map_err(|_| crate::ParseError::new::<PrefixOrderDomain>(value.clone()))?;
        Ok(PrefixOrderDomain { name })
    }
}

impl TryFrom<proto::Signature> for Unsigned {
    type Error = crate::ParseError;

    fn try_from(value: proto::Signature) -> Result<Self, Self::Error> {
        Ok(Unsigned {
            public_key: value.public_key,
        })
    }
}

impl From<Unsigned> for proto::Signature {
    fn from(unsigned: Unsigned) -> Self {
        proto::Signature::unsigned(unsigned.public_key)
    }
}

impl From<KeyPair> for Unsigned {
    fn from(value: KeyPair) -> Self {
        Unsigned {
            public_key: value.public_key().into(),
        }
    }
}

impl TryFrom<proto::Action> for Action {
    type Error = crate::ParseError;

    fn try_from(value: proto::Action) -> Result<Self, Self::Error> {
        match value.action {
            Some(proto::action::Action::Reconfigure(reconfigure)) => {
                Ok(Action::Reconfigure(reconfigure.try_into()?))
            }
            Some(proto::action::Action::Observe(observe)) => {
                Ok(Action::Observe(observe.try_into()?))
            }
            None => Err(crate::ParseError::new::<Action>("missing".to_string())),
        }
    }
}

impl From<Action> for proto::Action {
    fn from(action: Action) -> Self {
        match action {
            Action::Reconfigure(reconfigure) => proto::Action {
                action: Some(proto::action::Action::Reconfigure(reconfigure.into())),
            },
            Action::Observe(observe) => proto::Action {
                action: Some(proto::action::Action::Observe(observe.into())),
            },
        }
    }
}

impl TryFrom<proto::action::Reconfigure> for Reconfigure {
    type Error = crate::ParseError;

    fn try_from(value: proto::action::Reconfigure) -> Result<Self, Self::Error> {
        let proto::action::Reconfigure {
            signature,
            config,
            not_before,
            not_after,
        } = value;

        let admin: Admin = signature
            .map(Unsigned::try_from)
            .ok_or_else(|| crate::ParseError::new::<Admin>("missing".to_string()))??
            .try_into()?;

        let config = config
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<Config>("missing".to_string()))??;

        let not_before = not_before
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<tendermint::Time>("missing".to_string()))?
            .map_err(|_| crate::ParseError::new::<tendermint::Time>("missing".to_string()))?;

        let not_after = not_after
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<tendermint::Time>("missing".to_string()))?
            .map_err(|_| crate::ParseError::new::<tendermint::Time>("missing".to_string()))?;

        Ok(Reconfigure {
            admin,
            config,
            not_before,
            not_after,
        })
    }
}

impl From<Reconfigure> for proto::action::Reconfigure {
    fn from(reconfigure: Reconfigure) -> Self {
        let Reconfigure {
            admin,
            config,
            not_before,
            not_after,
        } = reconfigure;
        proto::action::Reconfigure {
            signature: Some(admin.into()),
            config: Some(config.into()),
            not_before: Some(not_before.into()),
            not_after: Some(not_after.into()),
        }
    }
}

impl TryFrom<proto::action::Observe> for Observe {
    type Error = crate::ParseError;

    fn try_from(value: proto::action::Observe) -> Result<Self, Self::Error> {
        let proto::action::Observe {
            signature,
            observation,
        } = value;

        let oracle = signature
            .map(Unsigned::try_from)
            .ok_or_else(|| crate::ParseError::new::<Oracle>("missing".to_string()))??
            .try_into()?;

        let observation = observation
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<Observation>("missing".to_string()))??;

        Ok(Observe {
            oracle,
            observation,
        })
    }
}

impl From<Observe> for proto::action::Observe {
    fn from(observe: Observe) -> Self {
        let Observe {
            oracle,
            observation,
        } = observe;
        proto::action::Observe {
            signature: Some(oracle.into()),
            observation: Some(observation.into()),
        }
    }
}

impl TryFrom<proto::Config> for Config {
    type Error = crate::ParseError;

    fn try_from(value: proto::Config) -> Result<Self, Self::Error> {
        let proto::Config {
            version,
            admin_config,
            oracle_config,
            onion_config,
        } = value;

        let version = version
            .try_into()
            .map_err(|_| crate::ParseError::new::<u32>("missing".to_string()))?;

        let admin_config = admin_config
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<AdminConfig>("missing".to_string()))??;

        let oracle_config = oracle_config
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<OracleConfig>("missing".to_string()))??;

        let onion_config = onion_config
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<OnionConfig>("missing".to_string()))??;

        Ok(Config {
            version,
            admins: admin_config,
            oracles: oracle_config,
            onion: onion_config,
        })
    }
}

impl From<Config> for proto::Config {
    fn from(config: Config) -> Self {
        let Config {
            version,
            admins: admin_config,
            oracles: oracle_config,
            onion: onion_config,
        } = config;
        proto::Config {
            version: version.into(),
            admin_config: Some(admin_config.into()),
            oracle_config: Some(oracle_config.into()),
            onion_config: Some(onion_config.into()),
        }
    }
}

impl TryFrom<proto::config::AdminConfig> for AdminConfig {
    type Error = crate::ParseError;

    fn try_from(value: proto::config::AdminConfig) -> Result<Self, Self::Error> {
        let proto::config::AdminConfig {
            admins,
            voting_config,
        } = value;

        let admins = admins
            .into_iter()
            .map(TryInto::try_into)
            .collect::<Result<_, _>>()?;

        let voting_config = voting_config
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<VotingConfig>("missing".to_string()))??;

        Ok(AdminConfig {
            authorized: admins,
            voting: voting_config,
        })
    }
}

impl From<AdminConfig> for proto::config::AdminConfig {
    fn from(config: AdminConfig) -> Self {
        let AdminConfig {
            authorized: admins,
            voting: voting_config,
        } = config;
        proto::config::AdminConfig {
            admins: admins.into_iter().map(Into::into).collect(),
            voting_config: Some(voting_config.into()),
        }
    }
}

impl TryFrom<proto::config::OracleConfig> for OracleConfig {
    type Error = crate::ParseError;

    fn try_from(value: proto::config::OracleConfig) -> Result<Self, Self::Error> {
        let proto::config::OracleConfig {
            enabled,
            oracles,
            voting_config,
            max_enrolled_subdomains,
            observation_timeout,
        } = value;

        let max_enrolled_subdomains: u64 = u64::try_from(max_enrolled_subdomains)
            .map_err(|_| crate::ParseError::new::<u64>("missing".to_string()))?;

        let oracles = oracles
            .into_iter()
            .map(TryInto::try_into)
            .collect::<Result<_, _>>()?;

        let voting_config = voting_config
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<VotingConfig>("missing".to_string()))??;

        let observation_timeout = Duration::from_secs(
            u64::try_from(observation_timeout)
                .map_err(|_| crate::ParseError::new::<Duration>("missing".to_string()))?,
        );

        Ok(OracleConfig {
            enabled,
            authorized: oracles,
            voting: voting_config,
            observation_timeout,
            max_enrolled_subdomains,
        })
    }
}

impl From<OracleConfig> for proto::config::OracleConfig {
    fn from(config: OracleConfig) -> Self {
        let OracleConfig {
            enabled,
            authorized: oracles,
            voting: voting_config,
            max_enrolled_subdomains,
            observation_timeout,
        } = config;
        proto::config::OracleConfig {
            enabled,
            oracles: oracles.into_iter().map(Into::into).collect(),
            voting_config: Some(voting_config.into()),
            observation_timeout: i64::try_from(observation_timeout.as_secs()).unwrap_or(i64::MAX),
            max_enrolled_subdomains: i64::try_from(max_enrolled_subdomains).unwrap_or(i64::MAX),
        }
    }
}

impl TryFrom<proto::config::OnionConfig> for OnionConfig {
    type Error = crate::ParseError;

    fn try_from(value: proto::config::OnionConfig) -> Result<Self, Self::Error> {
        let proto::config::OnionConfig { enabled } = value;
        Ok(OnionConfig { enabled })
    }
}

impl From<OnionConfig> for proto::config::OnionConfig {
    fn from(config: OnionConfig) -> Self {
        let OnionConfig { enabled } = config;
        proto::config::OnionConfig { enabled }
    }
}

impl TryFrom<proto::config::VotingConfig> for VotingConfig {
    type Error = crate::ParseError;

    fn try_from(value: proto::config::VotingConfig) -> Result<Self, Self::Error> {
        let proto::config::VotingConfig {
            total,
            quorum,
            timeout,
            delay,
        } = value;

        let total =
            Total(u64::try_from(total).map_err(|_| crate::ParseError::new::<Total>(total))?);
        let quorum =
            Quorum(u64::try_from(quorum).map_err(|_| crate::ParseError::new::<Quorum>(quorum))?);
        let timeout = Timeout(Duration::from_secs(
            u64::try_from(timeout).map_err(|_| crate::ParseError::new::<Timeout>(timeout))?,
        ));
        let delay = Delay(Duration::from_secs(
            u64::try_from(delay).map_err(|_| crate::ParseError::new::<Delay>(delay))?,
        ));

        Ok(VotingConfig {
            total,
            quorum,
            timeout,
            delay,
        })
    }
}

impl From<VotingConfig> for proto::config::VotingConfig {
    fn from(config: VotingConfig) -> Self {
        let VotingConfig {
            total,
            quorum,
            timeout,
            delay,
        } = config;
        proto::config::VotingConfig {
            total: i64::try_from(total.0).unwrap_or(i64::MAX),
            quorum: i64::try_from(quorum.0).unwrap_or(i64::MAX),
            timeout: i64::try_from(timeout.0.as_secs()).unwrap_or(i64::MAX),
            delay: i64::try_from(delay.0.as_secs()).unwrap_or(i64::MAX),
        }
    }
}

impl TryFrom<proto::action::observe::Observation> for Observation {
    type Error = crate::ParseError;

    fn try_from(value: proto::action::observe::Observation) -> Result<Self, Self::Error> {
        let proto::action::observe::Observation {
            domain,
            zone,
            hash_observed,
            blockstamp,
        } = value;

        let domain = Domain {
            name: FQDN::from_ascii_str(&domain)
                .map_err(|_| crate::ParseError::new::<Domain>(domain))?,
        };

        let zone = Zone {
            name: FQDN::from_ascii_str(&zone).map_err(|_| crate::ParseError::new::<Zone>(zone))?,
        };

        let hash_observed = hash_observed
            .map(TryInto::try_into)
            .ok_or_else(|| crate::ParseError::new::<HashObserved>("missing".to_string()))??;

        let blockstamp = blockstamp
            .ok_or_else(|| crate::ParseError::new::<Blockstamp>("missing".to_string()))?
            .try_into()
            .map_err(|_| crate::ParseError::new::<Blockstamp>("missing".to_string()))?;

        Ok(Observation {
            domain,
            zone,
            hash_observed,
            blockstamp,
        })
    }
}

impl From<Observation> for proto::action::observe::Observation {
    fn from(observation: Observation) -> Self {
        let Observation {
            domain,
            zone,
            hash_observed,
            blockstamp,
        } = observation;
        proto::action::observe::Observation {
            domain: domain.into(),
            zone: zone.into(),
            hash_observed: Some(hash_observed.into()),
            blockstamp: Some(blockstamp.into()),
        }
    }
}

impl TryFrom<proto::action::observe::observation::Blockstamp> for Blockstamp {
    type Error = crate::ParseError;

    fn try_from(
        value: proto::action::observe::observation::Blockstamp,
    ) -> Result<Self, Self::Error> {
        let proto::action::observe::observation::Blockstamp {
            block_hash,
            block_number,
        } = value;

        let block_hash = block_hash
            .to_vec()
            .try_into()
            .map_err(|_| crate::ParseError::new::<AppHash>("missing".to_string()))?;

        let block_height: Height = u64::try_from(block_number)
            .map_err(|_| crate::ParseError::new::<Height>("missing".to_string()))?
            .try_into()
            .map_err(|_| crate::ParseError::new::<Height>("missing".to_string()))?;

        Ok(Blockstamp {
            app_hash: block_hash,
            block_height,
        })
    }
}

impl From<Blockstamp> for proto::action::observe::observation::Blockstamp {
    fn from(blockstamp: Blockstamp) -> Self {
        let Blockstamp {
            app_hash,
            block_height,
        } = blockstamp;
        proto::action::observe::observation::Blockstamp {
            block_hash: app_hash.as_bytes().to_vec().into(),
            block_number: i64::try_from(block_height.value()).unwrap_or(i64::MAX),
        }
    }
}

impl TryFrom<proto::Admin> for Admin {
    type Error = crate::ParseError;

    fn try_from(value: proto::Admin) -> Result<Self, Self::Error> {
        Ok(Admin {
            identity: value.public_key,
        })
    }
}

impl TryFrom<Unsigned> for Admin {
    type Error = crate::ParseError;

    fn try_from(value: Unsigned) -> Result<Self, Self::Error> {
        Ok(Admin {
            identity: value.public_key,
        })
    }
}

impl From<Admin> for proto::Admin {
    fn from(admin: Admin) -> Self {
        proto::Admin {
            public_key: admin.identity,
        }
    }
}

impl From<Admin> for proto::Signature {
    fn from(admin: Admin) -> Self {
        proto::Signature::unsigned(admin.identity)
    }
}

impl TryFrom<proto::Oracle> for Oracle {
    type Error = crate::ParseError;

    fn try_from(value: proto::Oracle) -> Result<Self, Self::Error> {
        Ok(Oracle {
            identity: value.public_key,
        })
    }
}

impl TryFrom<Unsigned> for Oracle {
    type Error = crate::ParseError;

    fn try_from(value: Unsigned) -> Result<Self, Self::Error> {
        Ok(Oracle {
            identity: value.public_key,
        })
    }
}

impl From<Oracle> for proto::Oracle {
    fn from(oracle: Oracle) -> Self {
        proto::Oracle {
            public_key: oracle.identity,
        }
    }
}

impl From<Oracle> for proto::Signature {
    fn from(oracle: Oracle) -> Self {
        proto::Signature::unsigned(oracle.identity)
    }
}

impl From<FQDN> for Domain {
    fn from(name: FQDN) -> Self {
        Domain { name }
    }
}

impl From<Domain> for FQDN {
    fn from(domain: Domain) -> Self {
        domain.name
    }
}

impl From<Empty> for String {
    fn from(_: Empty) -> Self {
        String::new()
    }
}

impl TryFrom<String> for Empty {
    type Error = crate::ParseError;

    fn try_from(string: String) -> Result<Self, crate::ParseError> {
        string
            .parse()
            .map_err(|_| crate::ParseError::new::<Empty>(string))
    }
}

impl TryFrom<proto::action::observe::observation::HashObserved> for HashObserved {
    type Error = crate::ParseError;

    fn try_from(
        value: proto::action::observe::observation::HashObserved,
    ) -> Result<Self, Self::Error> {
        if value.hash.len() == 32 {
            let mut hash = [0u8; 32];
            hash.copy_from_slice(&value.hash);
            Ok(HashObserved::Hash(hash))
        } else if value.hash.is_empty() {
            Ok(HashObserved::NotFound)
        } else {
            Err(crate::ParseError::new::<HashObserved>(hex::encode(
                value.hash,
            )))
        }
    }
}

impl From<HashObserved> for proto::action::observe::observation::HashObserved {
    fn from(value: HashObserved) -> Self {
        match value {
            HashObserved::Hash(hash) => proto::action::observe::observation::HashObserved {
                hash: hash.to_vec().into(),
            },
            HashObserved::NotFound => {
                proto::action::observe::observation::HashObserved { hash: Bytes::new() }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn domain_parse() {
        let domain_str = "com.";
        let domain: FQDN = domain_str.parse().unwrap();
        assert_eq!(domain.to_string(), domain_str);
    }
}
