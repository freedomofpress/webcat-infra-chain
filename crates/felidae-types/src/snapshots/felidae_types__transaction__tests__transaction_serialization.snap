---
source: crates/felidae-types/src/transaction.rs
expression: "serde_json::to_string(&tx).unwrap()"
---
{"chain_id":"test-chain","actions":[{"reconfigure":{"version":1,"admins":{"voting":{"total":3,"quorum":2,"timeout":"30s","delay":"5s"},"authorized":["00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"]},"oracles":{"enabled":true,"voting":{"total":5,"quorum":3,"timeout":"1m","delay":"10s"},"max_enrolled_subdomains":100,"observation_timeout":"2m","authorized":["01010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101"]},"onion":{"enabled":false},"not_before":"2022-04-15T05:20:00Z","not_after":"2022-08-08T23:06:40Z","admin":"00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"}}]}
